async_producer_test.go

```diff
@@ -2199,57 +2200,86 @@ func TestPreventRetryBufferOverflow(t *testing.T) {
-
-	producer, err := NewAsyncProducer([]string{broker.Addr()}, config)
-	if err != nil {
-		t.Fatal(err)
+	tests := []struct {
+		name            string
+		configureBuffer func(*Config)
+		messageSize     int
+		numMessages     int
+	}{
+		{
+			name: "MaxBufferLength",
+			configureBuffer: func(config *Config) {
+				config.Producer.Flush.MaxMessages = 1
+				config.Producer.Retry.MaxBufferLength = minFunctionalRetryBufferLength
+			},
+			messageSize: 1, // Small message size
+			numMessages: 10000,
+		},
+		{
+			name: "MaxBufferBytes",
+			configureBuffer: func(config *Config) {
+				config.Producer.Flush.MaxMessages = 1
+				config.Producer.Retry.MaxBufferBytes = minFunctionalRetryBufferBytes
+			},
+			messageSize: 950 * 1024, // 950 KB
+			numMessages: 1000,
+		},
 	}
 
-	var (
-		wg                        sync.WaitGroup
-		successes, producerErrors int
-		errorFound                bool
-	)
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for range producer.Successes() {
-			successes++
-		}
-	}()
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			config := NewTestConfig()
+			config.Producer.Return.Successes = true
+			tt.configureBuffer(config)
 
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for errMsg := range producer.Errors() {
-			if errors.Is(errMsg.Err, ErrProducerRetryBufferOverflow) {
-				errorFound = true
+			producer, err := NewAsyncProducer([]string{broker.Addr()}, config)
+			if err != nil {
+				t.Fatal(err)
 			}
-			producerErrors++
-		}
-	}()
 
-	numMessages := 100000
-	for i := 0; i < numMessages; i++ {
-		kv := StringEncoder(strconv.Itoa(i))
-		producer.Input() <- &ProducerMessage{
-			Topic:    topic,
-			Key:      kv,
-			Value:    kv,
-			Metadata: i,
-		}
-	}
+			var (
+				wg                        sync.WaitGroup
+				successes, producerErrors int
+				errorFound                bool
+			)
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for range producer.Successes() {
+					successes++
+				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for errMsg := range producer.Errors() {
+					if errors.Is(errMsg.Err, ErrProducerRetryBufferOverflow) {
+						errorFound = true
+					}
+					producerErrors++
+				}
+			}()
 
-	producer.AsyncClose()
-	wg.Wait()
+			longString := strings.Repeat("a", tt.messageSize)
+			val := StringEncoder(longString)
 
-	require.Equal(t, successes+producerErrors, numMessages, "Expected all messages to be processed")
-	require.True(t, errorFound, "Expected at least one error matching ErrProducerRetryBufferOverflow")
+			for i := 0; i < tt.numMessages; i++ {
+				msg := &ProducerMessage{
+					Topic: topic,
+					Value: val,
+				}
+				producer.Input() <- msg
+			}
+
+			producer.AsyncClose()
+			wg.Wait()
+
+			require.Equal(t, successes+producerErrors, tt.numMessages, "Expected all messages to be processed")
+			require.True(t, errorFound, "Expected at least one error matching ErrProducerRetryBufferOverflow")
```

Comment:
There’s not really a reason why these should be `require` instead of `assert`. https://go.dev/wiki/TestComments#keep-going

Reply:
Thank you for your feedback! It helped me learn new things, and I’ve updated my code accordingly.